### PR TITLE
Add memory barriers to native AOT asm helpers

### LINKED ISSUE #81151: [NativeAOT] ordering issue around ensuring static constructor execution
Re: https://github.com/dotnet/runtime/blob/983c8f239a98812498d874ed36b7001bd764fdfe/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/CompilerServices/ClassConstructorRunner.cs#L59

We check `pContext->initialized` with an ordinary read, before accessing the static fields. There is a possibility that we will read the fields before reading the flag on weak-ordered architectures.

The code orders the write to the flag after running constructors[ by doing full memory fence](https://github.com/dotnet/runtime/blob/983c8f239a98812498d874ed36b7001bd764fdfe/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/CompilerServices/ClassConstructorRunner.cs#L95), but that is both too much and not sufficient. 
Too much is because just doing `Volatile.Write(ref pContext->initialized, 1)` is enough to guarantee the order of writes.
Not sufficient because the read side can still read ahead of the flag check.

A simple fix would be to do volatile read of the flag when checking. 
However the same flag check is emitted by the JIT. Not sure if we emit `ldar` there. I suspect we do not. If that is a case, emitting `ldar` would be a simple solution.

If ultimately, we do not want to emit `ldar`, an alternative could be using a slightly different pattern where `initialized` instead of being a condition flag, could be a pointer used to access the statics. I.E. if it is `null` the cctor has not run yet, and if it is not `null`, use it as a pointer (or an offset even if it is fixed, as long as it is used to compute the address) to access static state, then we could benefit from data dependency ordering the reads.

Replacing `Interlocked.MemoryBarrier()` with a  process-wide barrier would work too, but seems too heavy.

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #81151

### CODE REVIEW

**Comment by jkotas** [comment]
RiscV and Loongarch need the same fix. I will open a follow up issue for that.

**Review by VSadov — APPROVED** [review]
LGTM!

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/Target_ARM/ARMReadyToRunHelperNode.cs [hunk 1]
```diff
@@ -33,6 +33,7 @@ protected override void EmitCode(NodeFactory factory, ref ARMEmitter encoder, bo
                             encoder.EmitMOV(encoder.TargetRegister.Arg2, factory.TypeNonGCStaticsSymbol(target));
                             encoder.EmitLDR(encoder.TargetRegister.Arg3, encoder.TargetRegister.Arg2,
                                             (short)-NonGCStaticsNode.GetClassConstructorContextSize(factory.Target));
+                            encoder.EmitDMB();
```
Comment (SingleAccretion) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/Target_ARM/ARMReadyToRunHelperNode.cs:36]: Question: is there any benefit to having a fast path check in the regular (non-generic) "get static base" helpers, considering they are always expanded in RyuJit?

Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/Target_ARM/ARMReadyToRunHelperNode.cs:36]: I agree that there is no reason for having the check in the non-generic static helpers. I am going to delete them.

**Review by VSadov — APPROVED** [review]

**Comment by jkotas** [comment]
There is actually a lot more simplification of the R2R asm helpers possible. I am going to submit it in separate PR and keep this one for the memory barriers only.

**Comment by jkotas** [comment]
> RiscV and Loongarch need the same fix. I will open a follow up issue for that.

Opened #106219

**Pull request merged**